### PR TITLE
feat: deep linking implementation for ToolHive Studio

### LINKED ISSUE #956: [Story] Deep Linking for ToolHive Studio
### Story Title

As a user, I want to open a specific server install page by clicking on a link

### Description

# Deep Linking Implementation for ToolHive Studio

Enable one-click navigation from error messages or documentation directly to ToolHive Studio's server installation page, providing a seamless UX when handling missing secrets or configuration issues.

### Use Case

When an MCP server reports a missing secret or configuration issue, provide a clickable link that:

1. Opens ToolHive Studio (or focuses if already open)
2. Navigates to the correct server installation page
3. Opens the installation dialog automatically

### UX Improvement

**Before:** User sees error → Must manually open Studio → Navigate to Registry → Search for server → Click install (5 steps)  
**After:** User sees error → Clicks link → **Done!** (1 click) ✨

---

### Custom URL Protocol (`toolhive://`)

* Registered on all platforms (macOS, Windows, Linux)
* **Supported Action:**
  * `install-server` - Navigate to server installation page with auto-opening dialog

**Example:**
```
toolhive://install-server?server=github&registry=official
```

### Automatic Dialog Opening

When navigating via deep link, the registry page automatically:
1. Navigates to the correct server's registry page
2. Opens the installation dialog
3. User can immediately fill in configuration and install

### Acceptance Criteria

1. **Deep link from terminal:**
   ```bash
   open "toolhive://install-server?server=github"
   ```

2. **HTTP control endpoint:**
   ```bash
   curl -X POST http://127.0.0.1:51234/navigate \
     -H "Content-Type: application/json" \
     -d '{"url": "toolhive://install-server?server=github"}'
   ```

3. **App not running:** Deep link launches app and navigates
4. **App already running:** Deep link focuses app and navigates

### Target Release

v0.0.0

### Priority

Medium

### Additional Context

meta-mcp doc https://docs.google.com/document/d/1N8ZEnc-t2pkvOYGGmS61Byu_1mqgqRe5M52Eh-CfAd4/edit?tab=t.0

https://github.com/StacklokLabs/research/issues/80
https://github.com/StacklokLabs/research/issues/79

### PR DESCRIPTION
**Pull request by therealnb**
# Deep Linking Implementation for ToolHive Studio
Closes #956 

## 🎯 Goal

Enable one-click navigation from error messages or documentation directly to ToolHive Studio's server installation page, providing a seamless UX when handling missing secrets or configuration issues.

### Use Case

When an MCP server reports a missing secret or configuration issue, provide a clickable link that:

1. Opens ToolHive Studio (or focuses if already open)
2. Navigates to the correct server installation page
3. Opens the installation dialog automatically

### UX Improvement

**Before:** User sees error → Must manually open Studio → Navigate to Registry → Search for server → Click install (5 steps)  
**After:** User sees error → Clicks link → **Done!** (1 click) ✨

---

#### Core Implementation

* **`main/src/deep-link.ts`** - Core deep linking logic  
  * Protocol handler (`toolhive://` URL scheme)
  * URL parsing with support for server names and registries
  * Deep link generation utilities  
  * Navigation handling logic
* **`main/src/control-server.ts`** - HTTP control endpoint for programmatic navigation  
  * Runs on `http://127.0.0.1:51234`
  * `GET /health` - Health check endpoint  
  * `POST /navigate` - Navigate with JSON body  
  * `GET /navigate?url=<encoded>` - Navigate with query parameter  
  * Graceful handling when port is already in use
* **`renderer/src/common/hooks/use-deep-link.tsx`** - React hook for deep link navigation  
  * Listens for deep link events from main process  
  * Handles router navigation to server installation page
  * Automatically opens installation dialog

#### Testing & Documentation

* **`main/src/tests/deep-link.test.ts`** - Comprehensive unit tests  
  * URL parsing tests  
  * Link generation tests  
  * Edge case handling (8 tests, all passing)
* **`docs/deep-linking.md`** - Complete user documentation  
  * Protocol specification  
  * Usage examples  
  * Integration guide for MCP server authors  
  * HTTP control API reference

#### `main/src/main.ts`

**Changes:**

* Import deep link functions (`handleDeepLink`, `registerProtocol`, `generateInstallServerLink`, `generateCliCommand`)
* Register `toolhive://` protocol on app startup
* Start HTTP control server when app is ready
* Stop control server on app quit and during shutdown
* Add `open-url` event handler (macOS deep link support)
* Add `second-instance` event handler (Windows/Linux deep link support)
* Add IPC handlers:  
  * `deep-link:generate-install-link` - Generate deep link URLs  
  * `deep-link:generate-cli-command` - Generate equivalent CLI commands
* Fixed Vite compatibility: use `import.meta.env` instead of `process.env` for Sentry DSN

#### `preload/src/preload.ts`

**Changes:**

* Expose deep link API to renderer process
* Add `deepLink.onNavigate()` listener registration
* Add `deepLink.generateInstallLink()` method
* Add `deepLink.generateCliCommand()` method

#### `renderer/src/routes/__root.tsx`

**Changes:**

* Import and initialize `useDeepLink` hook at the root level
* Ensures deep link handling works throughout the app

#### `renderer/src/routes/(registry)/registry_.$name.tsx`

**Changes:**

* Enhanced to automatically open install dialog when navigating via deep link
* Checks for `server` search parameter in URL
* Preserves existing functionality while adding deep link support

#### Configuration Files

* **`.gitignore`** - Added `.vscode/` to ignore VS Code settings
* **`knip.ts`** - Added `ignoreExportsUsedInFile` config for test files
* **`vitest.setup.ts`** - Added mocks for `electron` and `electron-log` to fix test execution

---

### 1. Custom URL Protocol (`toolhive://`)

* Registered on all platforms (macOS, Windows, Linux)
* **Supported Action:**
  * `install-server` - Navigate to server installation page with auto-opening dialog

**Example:**
```
toolhive://install-server?server=github&registry=official
```

### 2. HTTP Control Endpoint

When ToolHive Studio is running, a local HTTP server on `http://127.0.0.1:51234` provides programmatic navigation:

**Endpoints:**
* `GET /health` - Check if control server is running
* `POST /navigate` - Navigate using JSON body `{"url": "toolhive://..."}`
* `GET /navigate?url=<encoded-url>` - Navigate using query parameter

**Example:**
```bash
curl -X POST http://127.0.0.1:51234/navigate \
  -H "Content-Type: application/json" \
  -d '{"url": "toolhive://install-server?server=github"}'
```

### 3. Link & CLI Command Generation

Helper functions to generate deep links and equivalent CLI commands:

```typescript
// Generate deep link
const link = await window.electronAPI.deepLink.generateInstallLink('github', 'official')
// Returns: "toolhive://install-server?server=github&registry=official"

// Generate CLI command
const cmd = await window.electronAPI.deepLink.generateCliCommand('github', 'official')
// Returns: "thv run --registry official github"
```

### 4. Automatic Dialog Opening

When navigating via deep link, the registry page automatically:
1. Navigates to the correct server's registry page
2. Opens the installation dialog
3. User can immediately fill in configuration and install

---

## 🧪 Testing
1. **Deep link from terminal:**
   ```bash
   open "toolhive://install-server?server=github"
   ```

2. **HTTP control endpoint:**
   ```bash
   curl -X POST http://127.0.0.1:51234/navigate \
     -H "Content-Type: application/json" \
     -d '{"url": "toolhive://install-server?server=github"}'
   ```

3. **App not running:** Deep link launches app and navigates
4. **App already running:** Deep link focuses app and navigates

---

## 🎨 User Experience

The implementation focuses on simplicity and reliability:

1. **Minimal friction:** From error message to installation dialog in one click
2. **Automatic dialog opening:** No need to search or navigate manually
3. **Fallback to HTTP:** If custom protocol fails, HTTP endpoint provides alternative
4. **CLI command generation:** For users who prefer command-line workflows

**Result: 5 manual steps → 1 click!** ✨

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      "name": "Debug ToolHive Studio Main Process",
+      "type": "node",
+      "request": "launch",
+      "program": "${workspaceFolder}/node_modules/.bin/electron-forge",
+      "args": ["start", "--", "--inspect=9229"],
+      "env": {
+        "NODE_ENV": "development",
+        "PATH": "/Users/nigel/.nvm/versions/node/v22.20.0/bin:${env:PATH}"
+      },
+      "runtimeExecutable": "/Users/nigel/.nvm/versions/node/v22.20.0/bin/node",
```
[the remark this answers is not in the document]
Comment (therealnb) [reply .vscode/launch.json]: I removed that code

File: main/src/control-server.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+    testReq.on('error', () => {
+      resolve(false)
+    })
+
+    testReq.setTimeout(1000, () => {
+      testReq.destroy()
+      resolve(false)
+    })
+  })
```
[the remark this answers is not in the document]
Comment (therealnb) [reply main/src/control-server.ts:230]: This is fine - it is not called regularly.

**Comment by samuv** [comment]
closing this in favor of the implementation that @kantord  is working on
